lib-extra/src/main/java/com/diffplug/spotless/extra/config/EclipseConfiguration.java

```diff
@@ -0,0 +1,243 @@
+				version = newVersion;
+			} else {
+				throw new IllegalArgumentException(
+						String.format("Version '%s' is not part of the supported versions '%s'.",
+								versionOrUrl,
+								supportedEclipseVersions.stream().map(v -> v.toString()).collect(Collectors.joining(", "))));
+			}
+		}
+	}
+
+	private static URL getDefaultCoordinatesUrl(URL eclipseConfigContext, SemanticVersion version) {
+		try {
+			Path filePath = Paths.get(eclipseConfigContext.getPath(), String.format("v%s%s",
+					version.toString(), MavenCoordinates.GRADLE_LOCKFILE_EXTENSION));
+			return new URL(eclipseConfigContext.getProtocol(), eclipseConfigContext.getHost(), filePath.toString());
+		} catch (MalformedURLException e) {
+			/*
+			 *  This exception must be prevented by the strict syntax checking of
+			 *  SemanticVersion and unit-tests of all allowed versions.
+			 */
+			throw new IllegalStateException(e);
+		}
+	}
+
+	/**
+	 * Modify default Maven dependencies configured for the formatter.
+	 * The default dependencies are located in {@link #ECLIPSE_FORMATTER_RESOURCES}.
+	 */
+	public void setDependencies(String... coordinates) {
+		coordinateAdaptations = new MavenCoordinates();
+		coordinateAdaptations.update(coordinates);
+	}
+
+	/** Set settings files containing Eclipse preferences */
+	public void setPreferences(Iterable<File> settingsFiles) {
+		this.settingsFiles = settingsFiles;
+	}
+
+	/** Creates the state of the configuration */
+	public EclipseConfiguration.State get() {
+		/*
+		 * The current use case is tailored for Gradle.
+		 * Gradle calls this method only once per execution
+		 * and compares the State with the one of a previous run
+		 * for incremental building.
+		 * Hence a lazy construction is not required.
+		 */
+		return new State(
+				jarProvisioner,
+				createMavenCoordinates(),
+				version,
+				settingsFiles);
+	}
+
+	private MavenCoordinates createMavenCoordinates() {
+		MavenCoordinates coordinates = new MavenCoordinates();
+		coordinates.add(defaultCoordinatesUrl);
+		coordinates.update(coordinateAdaptations);
+		return coordinates;
+	}
+
+	/**
+	 * State of Eclipse configuration items, providing functionality to derived information
+	 * based on the state.
+	 */
+	public static class State implements Serializable {
+		// Not used, only the serialization output is required to determine whether the object has changed
+		private static final long serialVersionUID = 1L;
+
+		private final FileSignature settingsFiles;
+		private final MavenCoordinates coordinates;
+		private final SemanticVersion version;
+
+		/*
+		 * Fields are transient because not needed to uniquely identify a Eclipse configuration state, and also because
+		 * Gradle only needs this class to be serializable so it can compare its members for incremental builds.
+		 */
+		private transient ClassLoader lazyClassLoader;
+		@SuppressFBWarnings("SE_TRANSIENT_FIELD_NOT_RESTORED")
+		private final transient Provisioner jarProvisioner;
```

Comment:
Normally, JarState is part of the serialized state:

https://github.com/diffplug/spotless/blob/de8ea62dcb3b407f404bb9c5e6b7744c50bb3f52/lib/src/main/java/com/diffplug/spotless/scala/ScalaFmtStep.java#L66-L72

This allows you to remove `coordinates` and `version` from the state because their effects are captured by the JarState (you can keep them too if you need them elsewhere).  JarState is more reliable, consider this:

- repo1 has one set of artifacts
- repo2 has the same names and versions, but the artifacts have been patched with different content

JarState captures this, because it hashes the content of the jars.  The current approach would miss this.

Reply:
Since the resolution of the jars take time, I wanted to be sure that it's only done when necessary. So what I basically want to avoid is a download in the following scenario:
1.  Developer modifies code
2.  Modification is not valid (tests, findbugs,...)
3. Formatter not executed due to prior errors
The download should not be triggered before it is assured that the Formatter will be executed. When I work on code I want to see quickly whether it works.

I am not sure when Gradle requests the state.

Reply:
Spotless (gradle or maven) requests the state at the last possible moment - not until someone actually calls `FormatterFunc::apply`.  Great care has been taken to put caching and lazy-evaluation all into the framework - adding second layers adds complexity without adding any performance.